typescript-requiring-type-checking.js

```diff
@@ -15,23 +15,23 @@ module.exports = {
         /*
```

Comment:
Discussion for recommended rule: `'@typescript-eslint/unbound-method': 'error'`

Reply:
Lots of interesting violations of [unbound-method](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/unbound-method.md) in the Web Portal codebase with the error "Avoid referencing unbound methods which may cause unintentional scoping of `this`":
1. [Using lodash](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=%2FOptimalTest.WebPortal%2Fsrc%2Fapp%2Fbars-and-lines-widget%2Fcomponents%2Fbars-and-lines-component%2Fbars-and-lines.component.ts&version=GBmain&line=234&lineEnd=235&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents): `items = _.uniqWith(selectionObj.selection, _.isEqual);`. I'm not familiar enough with lodash or scoping rules to know if this is a true problem or how to fix it.
2. [Passing an instance method as a callback](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=%2FOptimalTest.WebPortal%2Fsrc%2Fapp%2Fboxplot-widget-new%2Fservices%2Fboxplot-options-creator.service.ts&version=GBmain&line=534&lineEnd=535&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) to a sort function. `serie.data.sort(this.sortPoints);` The fix is to change `sortPoints` to `static` (it doesn't use `this`).
3. [Calling `removeEventListener` with an instance method](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=%2FOptimalTest.WebPortal%2Fsrc%2Fapp%2Fchart-widget-common%2Fservices%2Flabel-creator.service.ts&version=GBmain&line=12&lineEnd=13&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents): `xAxisLabels[i].removeEventListener('mouseout', this.hideTooltip);` [MDN has a detailed description of this behavior](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#the_value_of_this_within_the_handler). Is this code doing anything wrong?
4. [Test code which calls a helper function](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=%2FOptimalTest.WebPortal%2Fsrc%2Fapp%2Fcore%2Fservices%2Fstream-response-extraction-factory.service.spec.ts&version=GBmain&line=32&lineEnd=32&lineStartColumn=16&lineEndColumn=28&lineStyle=plain&_a=contents) defined inline which doesn't use `this`.
5. [Passing an instance method](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=%2FOptimalTest.WebPortal%2Fsrc%2Fapp%2Fhistogram-widget%2Fservices%2Fhistogram-data.service.ts&version=GBmain&line=76&lineEnd=77&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) as a callback when it [does use `this`](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=%2FOptimalTest.WebPortal%2Fsrc%2Fapp%2Fcore%2Fservices%2Fbase-http.service.ts&version=GBmain&line=20&lineEnd=21&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents). If `this` isn't passed correctly then this could be catching a bug.

It's clearly a gap in my own knowledge that I can't tell if these are correct or not but I wonder if the rule will help developers in the same boat as me to educate themselves or just bewilder/annoy them.

Reply:
Recently read a blog called [Don't use functions as callbacks unless they're designed for it](https://jakearchibald.com/2021/function-callback-risks/) that gives more examples that seem to be in favor of this rule. I know I've asked folks to use lambdas in cases where arguments don't line up or for binding a `this` context in SystemLink PRs.

Reply:
I agree, it's a good read. One thing that might help developers who hit this error would be if the error message guided them towards arrow functions. I think it's possible for us to customize the error message in our rule config, so if people agree that's a good idea then I can create a proposal that does that.

Reply:
We started discussing this at today's meeting but didn't complete. Most of these cases should be solved with arrow functions, but we should talk about `1` in more detail. Interesting note is that the code in `3` is actually wrong: `removeEventListener` is not removing the same callback that `addEventListener` added.

Reply:
Should we make `ignoreStatic` true for this rule? An example using Angular's reactive forms:
```
documentForm = new FormGroup({
    name: new FormControl('', [
        Validators.required,
        DocumentSettingsComponent.invalidCharacters
    ]),
    description: new FormControl('', [
        DocumentSettingsComponent.invalidCharacters
    ]),
    workspace: new FormControl('', Validators.required)
});
```
By default, this rule will error on the `Validators.required` reference (which is static and coming directly from Angular's API).

Reply:
We discussed in person. General agreement that we should `ignoreStatic` even though there are some cases where static methods might be using `this`. The annoyance of having to disable the rule for Carson's case outweighs the benefit in other cases.

Reply:
@rajsite this config needs to be changed to turn on `ignoreStatic`